qiskit_experiments/framework/experiment_data.py

```diff
@@ -726,14 +729,158 @@ def add_data(
             data = [data]
 
         # Directly add non-job data
-        with self._result_data.lock:
-            for datum in data:
-                if isinstance(datum, dict):
+        for datum in data:
+            if isinstance(datum, dict):
+                with self._result_data.lock:
                     self._result_data.append(datum)
-                elif isinstance(datum, Result):
-                    self._add_result_data(datum)
+            elif isinstance(datum, Result):
+                self._add_result_data(datum)
+            else:
+                raise TypeError(f"Invalid data type {type(datum)}.")
+        self.create_child_data()
+        self.init_children_data()
+
+    @property
+    def __retrive_self_attrs_as_dict(self) -> dict:
+
+        """
+        retrive data about self to bootstrap
+        """
+
+        return {
+            "backend": self.backend,
+            "tags": self.tags,
+            "auto_save": self.auto_save,
+            "service": self.service,
+            "provider": self.provider,
+            "backed_name": self.backend_name,
+            "notes": self.notes,
+            "start_datetime": self.start_datetime,
+            "verbose": self.verbose,
+            "source": self.source,
+            "share_level": self.share_level,
+            "experiment_type": self.experiment_type,
+        }
+
+    def create_child_data(self):  # pylint: disable=inconsistent-return-statements
+
+        """Bootstrap Experiment data containers
+
+        Returns:
+            self : return itself for method calling
+        """
+
+        if (component_metadata := self.metadata.get("component_metadata", None)) is None:
+            return
+
+        while (new_idx := len(self._child_data)) < len(component_metadata):
+            child_data = ExperimentData(**self.__retrive_self_attrs_as_dict)
+            # Add automatically generated component experiment metadata
+            try:
+                this_data = component_metadata[new_idx].copy()
+                child_data.metadata.update(this_data)
+            except (KeyError, IndexError):
+                pass
+            try:
+                component_type = self.metadata["component_types"][new_idx]
+                child_data.experiment_type = component_type
+            except (KeyError, IndexError):
+                pass
+            self.add_child_data(child_data)
+
+        return self
+
+    def init_children_data(self):  # pylint: disable=inconsistent-return-statements
```

Comment:
I think this can be protected method if it's automatically called from `create_child_data`. Public methods are meant to conform to our deprecation policy which we cannot easily change behavior of the methods.